doc/fastmail.1
.if n \{\
.	ds ct "
.\}
.if t \{\
.	ds ct \\(co
.\}
.TH FASTMAIL 1L "Elm Version 2.5" "USENET Community Trust"
.SH NAME
fastmail - quick batch mail interface to a single address
.SH SYNOPSIS
.B fastmail
[-b bcc-list] [-c cc-list] [-C comments] [-d] [-f fromname]
[-i msg-id] [-r replyto] [-R references] [-s subject] filename|- address-list
.br
.SH DESCRIPTION
.I Fastmail
is a low-level interface to the mail system that allows batch
processing of mail.  It's intended for mailing to very large
groups of people in a staggered fashion. 
.PP
The starting options are;
.TP 1.0i
.B "-b bcc-list"
This allows a list of people to receive blind-carbon copies, or BCCs, of
the message.  This list should be full email addresses.
.TP
.B "-c cc-list"
This allows a list of people to receive carbon copies, or CCs, of
the message.  This list should be full email addresses.
.TP
.B "-C comments"
This allows a Comments: line to be added to the RFC822 header with any
text you desire.
.TP
.B "-d"
Debug.  This is helpful for strange, unfriendly errors from
the program (etc).
.TP
.B "-f from"
This overrides the users name in the From: line, so that if
the user was x@y, and their name was MrX then the default 
From: line would be "From: x@y (MrX)".  Using "-f Joe" when
invoking this, though, would change it to "From: x@y (Joe)"
.TP
.B "-i msg-id"
This allows a message-id to which this message refers.
.TP
.B "-r replyto"
Occasionally, you might send mail but want the replies to go
to a different address (very common with mailing lists).  
There is a header for this purpose called "Reply-To:" which
can be utilized by using this starting option.  For example,
we could send mail with a reply-to to list-request by
using "-r list-request".  The header generated would then
be of the form "Reply-To: list-request".
.TP
.B "-R references"
This allows descriptive/reference text for this message.
.TP
.B "-s subject"
The subject of the message is specified by using 
this starting option.
.PP
Either the name of the file containing the message, or a - to indicate
usage of standard-in is required.
.SH EXAMPLE
Let's say we're user "big" on machine "big-vax" and we have a
shell script called 'batch-mail' that contains the following
lines:
.nf

   #
   # Batch Mail - batch mailing of a file to a LOT of users
   #  
   # Usage: batch-mail "from" "subject" filename

   sender_copy = $LOGIN
   replyto = "The-Mr-Big-list"

   fastmail -b $sender_copy -r $replyto -f "$1" -s "$2" $3 person1
   sleep 10
   fastmail -r $replyto -f "$1" -s "$2" $3 person2
   sleep 10
   fastmail -r $replyto -f "$1" -s "$2" $3 person3
   sleep 10
   fastmail -r $replyto -f "$1" -s "$2" $3 person4

   < etc >

with the invocation:

   batch-mail "Mr. Big" "Warning to all" warning.text

.fi
would mail a copy of the 'warning.text' file to person1, person2, 
person3, etc.  "$LOGIN" will also receive a copy of the first message
in the mail, \fIsilently\fR.  Each resultant message will include the headers:
.nf

    From: big-vax!big (Mr. Big)
    Subject: Warning to all
    Reply-To: The-Mr-Big-list

.fi
This program should turn out to be considerably
faster than the alternative methods of accomplishing this task.
.SH FILES
/usr/lib/sendmail       sendmail transport if available
.br
/bin/rmail              transport if no sendmail
.br
/tmp/fastmail.$$        temporary file 
.SH AUTHOR
Elm Development Group
.SH SEE\ ALSO
sendmail(1), rmail(1), elm(1L)
.SH BUG REPORTS TO
Bill Pemberton  flash@virginia.edu
.SH COPYRIGHTS
\fB\*(ct\fRCopyright 1988-1995 by The USENET Community Trust
.br
Derived from Elm 2.0, \fB\*(ct\fR Copyright 1986, 1987 by Dave Taylor


Changes
Changes in 2.5 PL8

- fixed a hang when hitting certain 8-bit chars in the From line

- fixed a file descriptor leak (from Andreas Luik)

- removed all Expires: processing


Changes in 2.5 PL7

- a few display issues fixed (from Kenneth Reek)

- buffer overflow in frm fixed

Changes in 2.5 PL6

- Fixed message-id in the In-Reply-To header.

- Actually show the -r option in the usage

- Whacked a few more places where we might follow a symlink in tmpdir

- Fastmail didn't grok commas correctly.

- Change lock.c to not use errno unless there's an errno

Changes in 2.5 PL5

- A few more buffer overruns fixed.

Changes in 2.5 PL4

- little bug in aliasdb.c (from Scot W. Hetzel <scot@genroco.com>)

- Message-Id: could overflow (from Darin Wayrynen <darin@deru.net>)

- Fixed the permissions setting problem with the edit mailbox function

- Added a ResizeScreen() call after show_msg() calls to hopefully deal wit
  SIGWINCH problems (inspired from Yuke Zhuge <zhuge@stone.pdx.informix.com>)
  
- Fixed problems under SunOS 4.1.3 (from Paul J. Menchini <mench@mench.com>)

- Fixed content encoding (from David Woolley  <david@djwhome.demon.co.uk>)

- Attempt to stop crashes with long recipient list (from Gilles Detillieux 
  <grdetil@scrc.umanitoba.ca>)
  

Changes in 2.5 PL3

- Fixed the declaration of outbound_from_name (from Richard Jackson 
  <rjackson@gmu.edu>)

- Fixed the leapyear macro (from Gerhard Niklasch)
  
- Fixed editing of a long CC line (from Derek Godfrey <djg@sjc.ascend.com>)

- Mail that lands in the sentmail with attachments now has the 
  Content-disposition header text added to the message (from Dave 
  Spensley)
  
- Made allto a bit longer (suggested by Dave Spensley 
  <davesp@typhoon.xnet.com>)

- Fixed Configure to find (again) fcntl on Linux

- can_access will now allow you to write to a char device (/dev/null)

Changes in 2.5 PL2

- Fixed (again, maybe?) savebyalias from Adam Benjamin <aebenjam@opentext.com>

- Fixed bug were an alias delete would crash elm if the mailbox was empty

Changes in 2.5 PL1

- Fixed savebyalias from Michael D Hughes <mdh@logcabin.woods.bridge.com> 
  and Robert Schwartz <rjs@elt.com>
  
- Changes for 8 bit clean from H.J. Lu <hjl@varesearch.com>

Changes in 2.5 PL0

- One minor change to make an easier compile on SparcLinux.

Changes in 2.5 PL0pre8

- Y2K stuff from Ed Symanzik <zik@msu.edu>

Changes in 2.5 PL0pre7

- minor change to deal with egcs

Changes in 2.5 PL0pre6

- newalias nows does a geteuid() instead of getuid().  This is to lets 
  SysAdmins do "su <user> -c 'newalias'"
  

Changes in 2.5 PL0pre5

- Fix for GECOS entries with comma-delimited sub-fields from Chip 
  Rosenthal <chip@unicom.com>.

- Fix? for keys getting screw up after a ^Z from Chip Rosenthal
  <chip@unicom.com>.


Changes in 2.5 PL0pre4

- A few changes for BSD from Chip Rosenthal <chip@unicom.com>.

- A few changes to Configure suggested by Chip Rosenthal <chip@unicom.com>.


Changes in 2.5 PL0pre3

- Configure didn't see fcntl locking on Linux.  Fix from Sean McDermott 
  <spm@intel.com>


Changes in 2.5 beta 3

- Fixed a problem where elm would try to close a file twice when
  sending an attachment (from reichler@ews.uiuc.edu)

- Added docs for environment variables to the man page (from Eric
  Haszlakiewicz <nimenees@uaccess.net>)

Changes in 2.5 beta 2

- IncomingFolders no longer set Magic Mode on for everything

- Configure should now deal with Linux.

- Make //pattern honor current limit (from Keith Neufeld<neufeld@pvi.org>)

- year 2000 fix.....?

- Building on Solaris requires /usr/ccs/bin to be in the PATH

Changes in 2.5 beta 1

- b)ounce honor DONT_ADD_FROM (from Keith Neufeld<neufeld@pvi.org>)

- removed [encode] display and reply (from Keith Neufeld<neufeld@pvi.org>)

- make right/left arrows work in the file browser the same as in the
  message index. (from Keith Neufeld<neufeld@pvi.org>)

- fixed a bug in the incomingfolders code (it wouldn't save) (from Bob Schwartz
  <rjs@elt.com>)
  
- fixed the meaning of noheader for forwards (from Keith Neufeld<neufeld@pvi.org>)



Changes in 2.5 alpha 11

- fixed a bug in the alias code (thanks to Keith Neufeld<neufeld@pvi.org>)

- fixed the generation of temp file names in newmbox.c

- allowed longer Cc: lines.  This was due to a crash with a group reply.

- added incomingfolders to the elmrc.  This is a list of folders that will
  automatically be opened with "magic mode" on. (inspired by a patch from
  Cristophe Kalt <kalt@ensta.fr>)


Changes in 2.5 alpha 10

- Press u on a message that is read but not deleted will mark the message as
  unread (from Larry Schwimmer <rosebud@cyclone.Stanford.EDU>).

- mmencode.1 man page is now installed with others (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- Magic mode stuff is now in help and NLS (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- allow elimination of aliases from larger aliases (To: bigalias -subalias).
  Based on incomplete patch by Waldi Ravens <waldi@moacs.indiv.nl.net>.  Also
  removes redundant declarations of get_alias_address(), which is declared in
  elm_globals.h (from Keith Neufeld<neufeld@pvi.org>)

- On an entry field that normally clears to be retyped from scratch as soon
  as the user presses a key (Subject: on header edit screen), allow right
  arrow and delete to begin editing the existing default answer. (from Keith
  Neufeld<neufeld@pvi.org>)

- Add a warning to Configure about disabling shell escapes (from Keith
  Neufeld<neufeld@pvi.org>)

- Eliminate complaints in doc/Makefile about duplicate makefile entries if
  mansrc and catmansrc are both "none" (from Keith Neufeld<neufeld@pvi.org>)

- Allow entry of uppercase commands on header display/edit screens, since
  that's how they're parenthesized on the display.  (from Bob Schwartz
  <rjs@elt.com> and Keith Neufeld <neufeld@pvi.org>

- Make 2+ -line error messages write over more of the screen instead of

utils/fastmail.c


/*******************************************************************************
 *  The Elm Mail System  -  $Revision: 1.4 $   $State: Exp $
 *
 *                      Copyright (c) 1988-1995 USENET Community Trust
 * 			Copyright (c) 1986,1987 Dave Taylor
 *******************************************************************************
 * Bug reports, patches, comments, suggestions should be sent to:
 *
 *      Bill Pemberton, Elm Coordinator
 *      flash@virginia.edu
 *
 *******************************************************************************
 * $Log: fastmail.c,v $
 * Revision 1.4  1996/03/14  17:30:07  wfp5p
 * Alpha 9
 *
 * Revision 1.2  1995/04/20  21:02:06  wfp5p
 * Added the showreply feature and emacs key bindings.
 *
 * Revision 1.1.1.1  1995/04/19  20:38:40  wfp5p
 * Initial import of elm 2.4 PL0 as base for elm 2.5.
 *
 ******************************************************************************/

/** This program is specifically written for group mailing lists and
    such batch type mail processing.  It does NOT use aliases at all,
    it does NOT read the /etc/password file to find the From: name
    of the user and does NOT expand any addresses.  It is meant 
    purely as a front-end for either /bin/mail or /usr/lib/sendmail
    (according to what is available on the current system).

         **** This program should be used with CAUTION *****

**/

/** The calling sequence for this program is:

	fastmail {args}  [ filename | - ] full-email-address 

   where args could be any (or all) of;

	   -b bcc-list		(Blind carbon copies to)
	   -c cc-list		(carbon copies to)
	   -C comment-line      (Comments:)
	   -d			(debug on)
	   -f from 		(from name)
	   -F from-addr		(the actual address to be put in the From: line)
	   -i msg-id            (In-Reply-To: msgid)
	   -r reply-to-address 	(Reply-To:)
	   -R references        (References:)
	   -s subject 		(subject of message)
**/

#define INTERN
#include "elm_defs.h"
#include "patchlevel.h"
#include "s_fastmail.h"

int addword P_((char *, int, const char *, int *));

extern char *get_arpa_date();


main(argc, argv)
int argc;
char *argv[];
{
    extern char *optarg;
    extern int optind;
    FILE *ifp, *ofp;
    int fp_handle;
    char *progname, temp_fname[SLEN], cmdbuf[LONG_STRING], buf[LONG_STRING];
    char *subject, *cc_list, *bcc_list, to_list[LONG_STRING];
    char *from_addr, *from_fullname;
    char *replyto, *comments, *inreplyto, *references;
    char *p;
    int  add_from, i, len;

    initialize_common();
    progname = basename(argv[0]);
    sprintf(temp_fname, "%s/fastmail.%d", default_temp, getpid());
    subject = NULL;
    cc_list = NULL;
    bcc_list = NULL;
    to_list[0] = '\0';
    from_addr = NULL;
    from_fullname = NULL;
    replyto = getenv("REPLYTO");
    comments = NULL;
    inreplyto = NULL;
    references = NULL;
#ifdef DONT_ADD_FROM
    add_from = FALSE;
#else
    add_from = TRUE;
#endif

    while ((i = getopt(argc, argv, "b:c:C:df:F:i:r:R:s:")) != EOF) {
	switch (i) {
	case 'b' : bcc_list = optarg;				break;
	case 'c' : cc_list = optarg;				break;
	case 'C' : comments = optarg;				break;
#ifdef DEBUG
	case 'd' : debug++;					break;	
#else
	case 'd' :
	    fprintf(stderr, catgets(elm_msg_cat, FastmailSet, FastmailNoDebug,
			"%s: cannot use \"-d\" (not compiled with DEBUG)\n"),
			progname);
	    exit(1);
#endif
	case 'f' : from_fullname = optarg; add_from = TRUE;	break;
	case 'F' : from_addr = optarg;	 add_from = TRUE;	break;
	case 'i' : inreplyto = optarg;				break;
	case 'r' : replyto = optarg;				break;
	case 'R' : references = optarg;				break;
	case 's' : subject = optarg;				break;
	default  : goto usage_error;
	}
    }

    if (argc-optind < 2) {
usage_error:
	fprintf(stderr, catgets(elm_msg_cat, FastmailSet, FastmailUsage,
"Usage: %s [args ...] filename address ...\n\
\n\
\t\"filename\" may be \"-\" to send stdin\n\
\t\"args ...\" can be:\n\
\n\
\t-b bcc-list\n\
\t-c cc-list\n\
\t-C comments\n\
\t-f from-name\n\
\t-F from-addr\n\
\t-i msg-id\n\
\t-r reply-to\n\
\t-R references\n\
\t-s subject\n\
\t-d  (to enable debugging)\n\
\n"), progname);
	exit(1);
    }

    p = argv[optind++];
    if (streq(p, "-")) {
	ifp = stdin;
    } else if ((ifp = fopen(p, "r")) == NULL) {
	fprintf(stderr, catgets(elm_msg_cat, FastmailSet,
		    FastmailCannotOpenInput,
		    "%s: cannot open input file \"%s\" [%s]\n"),
		    progname, p, strerror(errno));
	exit(1);
    }

    for (len = 0 ; optind < argc ; ++optind) {
	i = strlen(argv[optind]);
	if (len+2+i >= sizeof(to_list)) {
	    fprintf(stderr, catgets(elm_msg_cat, FastmailSet,
			FastmailTooManyToRecip,
			"%s: too many \"To\" recipients\n"), progname);
	    exit(1);
	}
	if (len > 0) {
	    to_list[len++] = ',';
	    to_list[len++] = ' ';
	}
	strcpy(to_list+len, argv[optind]);
	len += i;
    }

    if (add_from) {
	if (!from_addr) {
#ifdef INTERNET
	    int l1 = strlen(user_name), l2 = strlen(host_fullname);
	    from_addr = (char *) safe_malloc(l1+1+l2+1);
	    sprintf(from_addr, "%s@%s", user_name, host_fullname);
#else
	    from_addr = user_name;
#endif
	}
	if (!from_fullname)
	    from_fullname = user_fullname;
    }

    (void) strcpy(cmdbuf, mailer);
    len = strlen(cmdbuf);
    if ((addword(cmdbuf, sizeof(cmdbuf), to_list, &len) < 0)
	|| (cc_list && addword(cmdbuf, sizeof(cmdbuf), cc_list, &len) < 0)
	|| (bcc_list && addword(cmdbuf, sizeof(cmdbuf), bcc_list, &len) < 0)
	|| (addword(cmdbuf, sizeof(cmdbuf), "<", &len) < 0)
	|| (addword(cmdbuf, sizeof(cmdbuf), temp_fname, &len) < 0)
    ) {
	fprintf(stderr, catgets(elm_msg_cat, FastmailSet, FastmailTooManyRecip,
		    "%s: too many recipients specified\n"), progname);
	exit(1);
    }

    
    
    if ( ((fp_handle = open(temp_fname, O_RDWR|O_CREAT|O_EXCL,0600)) != -1) &&
	 ((ofp = fdopen(fp_handle, "w")) == NULL) ) {
	fprintf(stderr, catgets(elm_msg_cat, FastmailSet,
		    FastmailCannotCreateTemp,
		    "%s: cannot create temp file %s [%s]\n"),
		    progname, temp_fname, strerror(errno));
	exit(1);
    }

    /*
     * Warning - the following comment is somewhat obsolete.
     * It does explain, however, why we do some seemingly
     * wierd things looking at the mailer before handling the subject.
     */

	/** Subject must appear even if "null" and must be first
	    at top of headers for mail because the
	    pure System V.3 mailer, in its infinite wisdom, now
	    assumes that anything the user sends is part of the 
	    message body unless either:
		1. the "-s" flag is used (although it doesn't seem
		   to be supported on all implementations?)
		2. the first line is "Subject:".  If so, then it'll
		   read until a blank line and assume all are meant
		   to be headers.
	    So the gory solution here is to move the Subject: line
	    up to the top.  I assume it won't break anyone elses program
	    or anything anyway (besides, RFC-822 specifies that the *order*
	    of headers is irrelevant).  Gahhhhh....

	    If we have been configured for a smart mailer then we don't want
	    to add a from line.  If the user has specified one then we have
	    to honor their wishes.  If they've just given a 'from name' then
	    we'll just put in the username and hope the mailer can add the
	    correct domain in.
	**/

    if (!subject) {
	i = strlen(mailer);
	if (i > 5 && streq(mailer+i-5, "/mail"))
	    subject = "";
	else
	if (i > 6 && streq(mailer+i-6, "/mailx"))
	    subject = "";
    }

    /*
     * Emit the message header.
     */
    if (subject)
	fprintf(ofp, "Subject: %s\n", subject);
    if (from_addr) {
	if (from_fullname)
	    fprintf(ofp, "From: %s (%s)\n", from_addr, from_fullname);
	else
	    fprintf(ofp, "From: %s\n", from_addr);
    }
    fprintf(ofp, "Date: %s\n", get_arpa_date());
    if (replyto)
	fprintf(ofp, "Reply-To: %s\n", replyto);
    fprintf(ofp, "To: %s\n", to_list);
    if (cc_list)
	fprintf(ofp, "Cc: %s\n", cc_list);
    if (references)
	fprintf(ofp, "References: %s\n", references);
    if (inreplyto)
	fprintf(ofp, "In-Reply-To: %s\n", inreplyto);
    if (comments)
	fprintf(ofp, "Comments: %s\n", comments);
#ifndef NO_XHEADER
    fprintf(ofp, "X-Mailer: fastmail [version %s PL%s]\n",
		VERSION, PATCHLEVEL);
#endif
    putc('\n', ofp);

    /*
     * Emit the message body.
     */
    while (fgets(buf, sizeof(buf), ifp) != NULL)
	fputs(buf, ofp);

    fclose(ifp);
    fclose(ofp);

    dprint(1, (debugfile, "executing: %s\n", cmdbuf));
   
    i = system(cmdbuf);
    (void) unlink(temp_fname);

    if (i != 0) {
	fprintf(stderr, catgets(elm_msg_cat, FastmailSet, FastmailErrorSending,
		"%s: error sending message [%s return status %d]\n"),
		progname, mailer, i);
	exit(1);
    }
    exit(i);
}


int addword(dest, destsiz, word, len_p)
char *dest;
int destsiz;
const char *word;
int *len_p;
{
    int dlen, wlen;

    dlen = *len_p;
    dest += dlen;
    wlen = strlen(word);
    if (dlen+1+wlen+1 > destsiz)
	return -1;

    if (dlen > 0) {
	*dest++ = ' ';
	++dlen;
    }

    for ( ; *word != '\0' ; ++word) {
	if (*word != ',') 
        {
	    *dest++ = *word;
	    ++dlen;
	}
        else
       {
	  *dest++ = ' ';
 	  ++dlen;
       }
       
    }
    *dest = '\0';

    *len_p = dlen;
    return 0;
}

